rclcpp/include/rclcpp/experimental/subscription_intra_process.hpp

```diff
@@ -163,8 +200,75 @@ class SubscriptionIntraProcess : public SubscriptionIntraProcessBase
     }
   }
 
+  template<typename T>
+  typename std::enable_if<
+    std::is_same<T, rcl_serialized_message_t>::value &&
+    std::is_same<MessageT, rclcpp::experimental::SerializedContainer>::value,
+    void>::type
+  execute_impl()
+  {
+    rmw_message_info_t msg_info;
+    msg_info.from_intra_process = true;
+
+    if (any_callback_.use_take_shared_method()) {
+      ConstMessageSharedPtr msg = buffer_->consume_shared();
+      if (msg == nullptr) {
+        throw std::runtime_error("Subscription intra-process could not get serialized message");
+      }
+      any_callback_.dispatch_intra_process(msg, msg_info);
+    } else {
+      throw std::runtime_error("Subscription intra-process for serialized "
+              "messages does not support unique pointers.");
+    }
+  }
+
+  template<class T>
+  typename std::enable_if<
+    !std::is_same<T, rcl_serialized_message_t>::value &&
+    std::is_same<MessageT, rclcpp::experimental::SerializedContainer>::value,
+    void>::type
+  execute_impl()
+  {
+    if (serializer_) {
+      ConstMessageSharedPtr serialized_container = buffer_->consume_shared();
+      if (serialized_container == nullptr) {
+        throw std::runtime_error("Subscription intra-process could not get serialized message");
+      }
+
+      rmw_message_info_t msg_info;
+      msg_info.from_intra_process = true;
+
+      if (any_callback_.use_take_shared_method()) {
+        CallbackMessageSharedPtr msg = construct_unique();
+        serializer_->deserialize_message(*serialized_container,
+          reinterpret_cast<void *>(msg.get()));
+        any_callback_.dispatch_intra_process(msg, msg_info);
+      } else {
+        CallbackMessageUniquePtr msg = construct_unique();
+        serializer_->deserialize_message(*serialized_container,
+          reinterpret_cast<void *>(msg.get()));
+        any_callback_.dispatch_intra_process(std::move(msg), msg_info);
+      }
+    } else {
+      throw std::runtime_error("Subscription intra-process can't handle unserialized messages");
+    }
+  }
+
+  CallbackMessageUniquePtr construct_unique()
```

Comment:
This could be a (inline) free function (anonymous namespace) which takes the allocator as a first argument.